### PR TITLE
Let DAGContext accepts Store instead of Snapshot

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

This PR lets `DAGContext` accepts `Store` instead of `Snapshot`. In this way we can do layered benchmark for `DAGContext` using a mocked `Store`.

Extracted from [batch execution PR](https://github.com/tikv/tikv/pull/3898)

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Existing unit tests.

## Does this PR affect documentation (docs) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

https://github.com/tikv/tikv/pull/3898

### CODE REVIEW

File: src/coprocessor/dag/pipeline.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    Executor, HashAggExecutor, IndexScanExecutor, LimitExecutor, SelectionExecutor,
+    StreamAggExecutor, TableScanExecutor, TopNExecutor,
+};
+use coprocessor::dag::expr::EvalConfig;
+use coprocessor::*;
+
+pub struct ExecutorPipelineBuilder;
+
+impl ExecutorPipelineBuilder {
+    pub fn build_normal<S: Store + 'static>(
```
Comment (rleungx) [inline src/coprocessor/dag/pipeline.rs]: We need some comments for these `pub` functions and traits.

File: src/coprocessor/dag/pipeline.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        execs: Vec<executor::Executor>,
```
Comment (rleungx) [inline src/coprocessor/dag/pipeline.rs]: How about `executors`?

File: src/coprocessor/dag/pipeline.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+use storage::Store;
+
+use super::executor::{
+    Executor, HashAggExecutor, IndexScanExecutor, LimitExecutor, SelectionExecutor,
+    StreamAggExecutor, TableScanExecutor, TopNExecutor,
+};
+use coprocessor::dag::expr::EvalConfig;
+use coprocessor::*;
+
+pub struct ExecutorPipelineBuilder;
```
Comment (huachaohuang) [inline src/coprocessor/dag/pipeline.rs]: What does the name "Pipeline" mean?

File: src/coprocessor/dag/pipeline.rs (hunk 1 above)
Comment (Connor1996) [reply src/coprocessor/dag/pipeline.rs]: +1

Comment (breeswish) [reply src/coprocessor/dag/pipeline.rs]: Thanks a lot for mentioning! I forgot adding it 🤣

File: src/coprocessor/dag/builder.rs [hunk 4]
```diff
@@ -25,22 +25,32 @@ use super::executor::{
 use coprocessor::dag::expr::EvalConfig;
 use coprocessor::*;
 
-pub struct ExecutorPipelineBuilder;
+/// Utilities to build an executor DAG.
+///
+/// Currently all executors are executed in sequence and there is no task schedulers, so this
+/// builder is in fact a pipeline builder. The builder will finally build a `Box<Executor>` which
+/// may contain another executor as its source, embedded in the field, one after another. These
+/// nested executors forms an executor pipeline that a single iteration at the out-most executor
```
Comment (rleungx) [inline src/coprocessor/dag/builder.rs]: form

File: src/coprocessor/dag/builder.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+/// (i.e. calling `next()`) will drive the whole pipeline.
+pub struct DAGBuilder;
+
+impl DAGBuilder {
+    /// Builds a normal executor pipeline.
+    ///
+    /// Normal executors iterate rows one by one.
```
[the remark this answers is not in the document]
Comment (breeswish) [reply src/coprocessor/dag/builder.rs]: there will be a `build_batch` later.

**Pull request merged**